firedrake/projection.py

```diff
@@ -35,72 +81,129 @@ def project(v, V, bcs=None, mesh=None,
-                                                   deg+1,
-                                                   shape=shape)
-        f = function.Function(fs)
-        f.interpolate(v)
-        v = f
-    elif isinstance(v, function.Function):
-        if v.function_space().mesh() != ret.function_space().mesh():
-            raise RuntimeError("Can't project between mismatching meshes")
-    elif not isinstance(v, ufl.core.expr.Expr):
-        raise RuntimeError("Can only project from expressions and functions, not %r" % type(v))
-
-    if v.ufl_shape != ret.ufl_shape:
-        raise RuntimeError('Shape mismatch between source %s and target function spaces %s in project' %
-                           (v.ufl_shape, ret.ufl_shape))
-
-    p = ufl_expr.TestFunction(V)
-    q = ufl_expr.TrialFunction(V)
-    a = ufl.inner(p, q) * ufl.dx(domain=V.mesh())
-    L = ufl.inner(p, v) * ufl.dx(domain=V.mesh())
-
-    # Default to 1e-8 relative tolerance
-    if solver_parameters is None:
-        solver_parameters = {'ksp_type': 'cg', 'ksp_rtol': 1e-8}
-    else:
-        solver_parameters.setdefault('ksp_type', 'cg')
-        solver_parameters.setdefault('ksp_rtol', 1e-8)
+            solver_parameters = solver_parameters.copy()
+        solver_parameters.setdefault("ksp_type", "cg")
+        self.source = source
+        self.target = target
+        self.solver_parameters = solver_parameters
+        self.form_compiler_parameters = form_compiler_parameters
+        self.bcs = bcs
+        self.constant_jacobian = constant_jacobian
+
+    @cached_property
+    def is_DG(self):
+        element = self.target.function_space().finat_element
+        return element.entity_closure_dofs() == element.entity_dofs()
+
+    @cached_property
+    def A(self):
+        u = firedrake.TrialFunction(self.target.function_space())
+        v = firedrake.TestFunction(self.target.function_space())
+        a = firedrake.inner(u, v)*firedrake.dx
+        if self.is_DG and self.bcs is None:
+            a = firedrake.Tensor(a).inv
+        A = firedrake.assemble(a, bcs=self.bcs,
+                               form_compiler_parameters=self.form_compiler_parameters)
+        return A
+
+    @cached_property
+    def solver(self):
+        if self.is_DG and self.bcs is None:
+            class Dummy(object):
+                def __init__(self, A):
+                    self.A = A
+
+                def solve(self, x, b):
+                    self.A.force_evaluation()
+                    with x.dat.vec_wo as x_, b.dat.vec_ro as b_:
+                        self.A.M.handle.mult(b_, x_)
+                    return x
+
+            return Dummy(self.A)
+        else:
+            solver = firedrake.LinearSolver(self.A, solver_parameters=self.solver_parameters)
+            return solver
+
+    @property
+    def apply_massinv(self):
+        if not self.constant_jacobian:
+            firedrake.assemble(self.A.a, tensor=self.A, bcs=self.bcs,
```

Comment:
I think this does one too many assemblies, because it makes `A` and then reassembles it.